### ISSUE TITLE
proposal: custom column name for edges

### ISSUE THREAD

**Issue by marwan-at-work** [facebookincubator/ent#126]
Currently there's a way to customize the Table name. Can we do the same for column names (including edges)? 

Thanks

**Comment by a8m** [comment]
There’s a [StorageKey](https://entgo.io/docs/schema-fields/#storage-key) option for columns. However, it does not support edges.

**Comment by marwan-at-work** [comment]
@a8m thank you! I will try it out. Is there a plan to support edges? It's the last piece of the puzzle to potentially adopt ent to an existing DB for my team :)

**Comment by a8m** [comment]
> Is there a plan to support edges? 

Yes. I was thinking about this when I added this option, but decided to ignore it because I wasn't 100% sure about the API. 

Basically, I thought about adding this option only to the `edge.To` builder (because `edge.From` is just a reference). Also, in addition to the `StorageKey(string)` option, add another method `StorageKeys(c1, c2 string)` for cases where the edge is mapped into a join table (in Gremlin, it always one storage-key).

**Comment by marwan-at-work** [comment]
@a8m that sounds great. Not sure I understand the two key vs one key though. I was imagining something on the edge itself where I can say `edge.To("cars").StorageKey("card_id")`. But whichever API lets me customize the column table of an edge I'm happy with for sure :+1: 

Thanks

**Comment by a8m** [comment]
> Not sure I understand the two key vs one key though.

Given this example:
```go
func (User) Edges() []ent.Edge {
    edge.To("friends", User.Type)
}
```
You have a many-to-many relation between a user to its friends. Therfore, a table named `user_friends` will be created with 2 columns (`user_id`, `friend_id`). These columns are the storage keys of this edge. 

Thanks for your feedback!

**Comment by clalos** [comment]
Hi guys, any update on this? We use custom column names for foreign keys in our MySQL db so I can't even try to query the relationships 😞

**Comment by a8m** [comment]
We still didn't figure out what is the best way for defining storage-key for edges (in terms of API design). Especially, in cases when you have M2M relationship and you want to control the table-name, FK1 and FK2. 

We open for suggestions.

**Comment by a8m** [comment]
> We open for suggestions.

I can also share my progress on this:
1. The `StorageKey` will be available only for `edge.To`
2. The definition form will be one of the 2 options (need to choose one): 
   ```go
   // M2M edge:
   edge.To("friends", User.Type).
       StorageKey(edge.StorageKey{Table: "friends", Columns: []string{"fk1", "fk2"}})

   // O2M edge:
   edge.To("pets", User.Type).
       StorageKey(edge.StorageKey{Columns: []string{"pet_owner"}})
   ```
   ```go
   // M2M edge:
   edge.To("friends", User.Type).
       StorageKey(edge.Table("friends"), edge.Columns("fk1", "fk2"))

   // O2M edge:
   edge.To("pets", User.Type).
       StorageKey(edge.Column("pet_owner"))
   ```

**Comment by clalos** [comment]
Looks fine. The second definition looks cleaner. I don't see any advantages in the more verbose one, maybe I'm missing something 🤔

**Comment by hyponymous** [comment]
Both APIs look fine to me. Two questions:

First question: for the M2M case, what's the interpretation of the order of `Columns`? Suppose we have
```
// in User
edge.To("groups", Group.Type).
	StorageKey(edge.StorageKey{
		Table: "my_rad_users_groups_junction_table",
		Columns: []string{"user_id", "group_id"},
	})

// in Group
edge.From("users", User.Type).
	Ref("groups")
```
Will it be correct to list `"user_id"` first here?

Since you're open to suggestions: given the potential ambiguity I'd suggest making these more explicit, e.g. with `FromColumn` and `ToColumn` fields. (Well-named arguments in the 2nd approach's `edge.Columns ` would also do it.)

Second question: projects moving onto `ent` will have pre-existing schemas and foreign key constraints with pre-existing names. Am I correct in understanding that `schema.ForeignKey`'s `Symbol` field represents the constraint name? if so, will I be able to define one (or two for M2M) in the edge?

Make that three questions: is there a plan to support additional columns on the M2M junction table?

**Comment by marwan-at-work** [comment]
@a8m any luck here? Second the last two comments in that both APIs look good to me :)

**Comment by afansv** [comment]
@a8m any updates? are there any related tasks i can help?

**Comment by afansv** [comment]
@a8m any updates? are there any related tasks i can help?

**Comment by a8m** [comment]
Thanks for the feedback @hyponymous. 

> Since you're open to suggestions: given the potential ambiguity I'd suggest making these more explicit, e.g. with FromColumn and ToColumn fields. (Well-named arguments in the 2nd approach's edge.Columns would also do it.)

I think `ToColumn/FromColumn` does make sense when defining column names.

> Am I correct in understanding that schema.ForeignKey's Symbol field represents the constraint name? if so, will I be able to define one (or two for M2M) in the edge?

Can you elaborate about your use case? We can do this, but it means one of the 2:
1. Skip ent foreign-key constraints creation, adopt the existing ones and just rely on them.
2. Same as 1, but verify the constraint details (like references and actions) on the migration. 

In my opinion, 1 doesn't make sense, because if there's a wrong constraint configuration we'll get unexpected behavior and I try to avoid this. The 2nd option looks like the right thing to do, but I don't think it's really worth the effort - users can drop the old constraints after the new ones were created.

**Comment by kosgug** [comment]
Any progress? I have the same doubts

**Comment by Vahanerevan** [comment]
+1

**Comment by a8m** [comment]
Please avoid commenting "+1" and use GitHub reactions. 

About the progress - I wanted to wait to get feedback for my [previous comment](https://github.com/facebookincubator/ent/issues/126#issuecomment-617313002). Since I didn't get any, I'll add this option to edges soon with option 1 as a "solution" for foreign-key constraints.

**Comment by marwan-at-work** [comment]
Solution 1 seems most ideal to me as well, because you most likely have a foreign key already set up if you were to specify a custom one. Looking forward to this feature!

**Comment by a8m** [comment]
Hey, I've created #545 for addressing this issue. Documentation will be added in a separate PR.

I decided to go with the following API:

```go
edge.To("pets", Pet.Type).
    StorageKey(edge.Column("owner_id"))

edge.To("groups", Group.Type).
    StorageKey(edge.Table("user_groups"), edge.Columns("user_id", "group_id"))
```

The first commit in the PR adopts the `ToColumn`/`FromColumn` API that [was proposed above](https://github.com/facebookincubator/ent/issues/126#issuecomment-568644015), but this was confusing, when to use `To`/`From`, because the destination of the join-column is an internal thing of entc and I don't expect users to know when the join-column is created on the inverse-edge table or the edge-owner table.

You're welcome to share you thoughts/review there. Thanks
